src/main/java/org/apache/solr/benchmarks/ControlledExecutor.java

```diff
@@ -65,181 +61,219 @@ public interface CallableWithType<V> extends Callable<V> {
         BenchmarksMain.OperationKey getType();
     }
 
-    public ControlledExecutor(String label, int threads, Integer duration, Integer rpm, Long maxExecution, int warmCount, Supplier<Callable<R>> taskSupplier) {
-        this(label, threads, duration, rpm, maxExecution, warmCount, taskSupplier, new ExecutionListener[0]);
+    public ControlledExecutor(String label, int threads, Integer duration, Integer rpm, Long maxExecution, int warmCount, Supplier<Callable<R>> actionSupplier) {
+        this(label, threads, duration, rpm, maxExecution, warmCount, actionSupplier, new ExecutionListener[0]);
     }
-    public ControlledExecutor(String label, int threads, Integer duration, Integer rpm, Long maxExecution, int warmCount, Supplier<Callable<R>> taskSupplier, ExecutionListener<BenchmarksMain.OperationKey, R>... executionListeners) {
-        this.label = label;
+    public ControlledExecutor(String taskName, int threads, Integer duration, Integer rpm, Long maxExecution, int warmCount, Supplier<Callable<R>> actionSupplier, ExecutionListener<BenchmarksMain.OperationKey, R>... executionListeners) {
+        this.taskName = taskName;
         this.duration = duration;
         this.maxExecution = maxExecution;
         this.warmCount = warmCount;
         this.stats = new SynchronizedDescriptiveStatistics();
-        this.taskSupplier = taskSupplier;
+        this.actionSupplier = actionSupplier;
         this.executionListeners = executionListeners;
+        this.maxPendingActions = threads * 10; //at most 10 * # of thread pending actions
+        this.rpm = rpm;
         executor = new ThreadPoolExecutor(threads, threads,
                 0L, TimeUnit.MILLISECONDS,
                 new LinkedBlockingQueue<>(),
-                new ThreadFactory() {
-                    @Override
-                    public Thread newThread(Runnable r) {
-                        return new Thread(r, ControlledExecutor.class.getSimpleName() + "-" + label);
-                    }
-                });
-        rateLimiter = rpm != null ? new RateLimiter(rpm) : null;
-
-        backPressureLimiter = new BackPressureLimiter(threads * 10); //at most 10 * # of thread pending tasks
+                r -> new Thread(r, ControlledExecutor.class.getSimpleName() + "-" + taskName));
     }
 
     public void run() throws InterruptedException, ExecutionException {
-        startTime = System.currentTimeMillis();
+        long startTime = System.currentTimeMillis();
+        Long endTime = duration != null ?  startTime + (1000 * duration) : null;
```

Comment:
Why is this a `Long` instead of a native `long`? I notice that it gets passed to `new StatusChecker()`, which also uses a `Long`, but it doesn't look like that field needs to be a `Long` either.

Reply:
This is for task that has no limit on duration (for example and indexing task, which it ends when all the input docs are exhausted), in such case we could use `null` which is only possible in wrapper class :)

Reply:
Ah of course, that makes sense!